src/state/config.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""gateway-api-integrator configuration."""
+
+import itertools
+import logging
+import typing
+
+import ops
+from lightkube import Client
+from lightkube.generic_resource import create_global_resource
+from pydantic import Field, ValidationError
+from pydantic.dataclasses import dataclass
+
+CUSTOM_RESOURCE_GROUP_NAME = "gateway.networking.k8s.io"
+GATEWAY_CLASS_RESOURCE_NAME = "GatewayClass"
+GATEWAY_CLASS_PLURAL = "gatewayclasses"
+
+logger = logging.getLogger()
+
+
+class InvalidCharmConfigError(Exception):
+    """Exception raised when a charm configuration is found to be invalid.
+
+    Attrs:
+        msg (str): Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the InvalidCharmConfigError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+@dataclass(frozen=True)
+class CharmConfig:
+    """A component of charm state that contains the charm's configuration.
+
+    Attrs:
+        gateway_class (_type_): _description_
+        external_hostname: The configured gateway hostname.
+    """
+
+    gateway_class: str = Field()
+    external_hostname: str = Field(
+        min_length=1, pattern=r"[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*"
+    )
+
+    @classmethod
+    def from_charm(cls, charm: ops.CharmBase) -> "CharmConfig":
+        """Create a CharmConfig class from a charm instance.
+
+        Args:
+            charm (ops.CharmBase): _description_
+
+        Raises:
+            InvalidCharmConfigError: _description_
+
+        Returns:
+            CharmConfig: _description_
+        """
+        gateway_class = typing.cast(str, charm.config.get("gateway-class"))
+        # my-py don't know that charm.client will always be initialized here
+        client: Client = charm.client  # type: ignore[attr-defined]
+        gateway_class_generic_resource = create_global_resource(
+            CUSTOM_RESOURCE_GROUP_NAME, "v1", GATEWAY_CLASS_RESOURCE_NAME, GATEWAY_CLASS_PLURAL
+        )
+        gateway_classes = tuple(client.list(gateway_class_generic_resource))
+        if not gateway_classes:
+            logger.error("No gateway class available on cluster.")
+            raise InvalidCharmConfigError("No gateway class available on cluster.")
```

Comment:
Is this really a config error? `gateway_classes` does not depend on the charm config.

Reply:
I was also not sure about this point but hesitated. I changed this to raise a more relevant error.